internal/services/containers/container_registry_cache_rule_resource.go

```diff
@@ -0,0 +1,277 @@
+		},
+		"container_registry_id": {
+			Type:         pluginsdk.TypeString,
+			Required:     true,
+			ForceNew:     true,
+			ValidateFunc: registries.ValidateRegistryID,
+		},
+		"credential_set_id": {
+			Type:         pluginsdk.TypeString,
+			Optional:     true,
+			Description:  "The ARM resource ID of the credential store which is associated with the cache rule.",
+			ValidateFunc: credentialsets.ValidateCredentialSetID,
+		},
+		"source_repo": {
+			Type:        pluginsdk.TypeString,
+			Required:    true,
+			ForceNew:    true,
+			Description: "The full source repository path such as 'docker.io/library/ubuntu'.",
+		},
+
+		"target_repo": {
+			Type:        pluginsdk.TypeString,
+			Required:    true,
+			ForceNew:    true,
+			Description: "The target repository namespace such as 'ubuntu'.",
+		},
+	}
+}
+
+func (ContainerRegistryCacheRule) Attributes() map[string]*pluginsdk.Schema {
+	return map[string]*pluginsdk.Schema{}
+}
+
+type ContainerRegistryCacheRuleModel struct {
+	Name                string `tfschema:"name"`
+	ContainerRegistryId string `tfschema:"container_registry_id"`
+	CredentialSetId     string `tfschema:"credential_set_id"`
+	SourceRepo          string `tfschema:"source_repo"`
+	TargetRepo          string `tfschema:"target_repo"`
+}
+
+func (ContainerRegistryCacheRule) ModelObject() interface{} {
+	return &ContainerRegistryCacheRuleModel{}
+}
+
+func (ContainerRegistryCacheRule) ResourceType() string {
+	return "azurerm_container_registry_cache_rule"
+}
+
+func (r ContainerRegistryCacheRule) Create() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 30 * time.Minute,
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			cacheRulesClient := metadata.Client.Containers.CacheRulesClient
+			subscriptionId := metadata.Client.Account.SubscriptionId
+			ctx, cancel := timeouts.ForCreate(metadata.Client.StopContext, metadata.ResourceData)
+
+			var config ContainerRegistryCacheRuleModel
+			if err := metadata.Decode(&config); err != nil {
+				return err
+			}
+
+			defer cancel()
+			log.Printf("[INFO] preparing arguments for Container Registry Cache Rule creation.")
+
+			registryId, err := registries.ParseRegistryID(metadata.ResourceData.Get("container_registry_id").(string))
+			if err != nil {
+				return err
+			}
+
+			id := cacherules.NewCacheRuleID(subscriptionId,
+				registryId.ResourceGroupName,
+				registryId.RegistryName,
+				metadata.ResourceData.Get("name").(string),
+			)
+
+			existing, err := cacheRulesClient.Get(ctx, id)
+			if err != nil {
+				if !response.WasNotFound(existing.HttpResponse) {
+					return fmt.Errorf("checking for presence of existing %s: %s", id, err)
+				}
+			}
+
+			if !response.WasNotFound(existing.HttpResponse) {
+				return tf.ImportAsExistsError("azurerm_container_registry_cache_rule", id.ID())
+			}
+
+			// TODO: make a check that the repo is available in the registry.
+			// TODO: validate the source repo.
```

Comment:
Do these still need to be done?

Reply:
I don't think so--issues with the source and target repo configuration will be caught in the `cacheRulesClient.CreateThenPoll` later in this function. We couldn't really use a validation function for this either, since it would involve making API calls against Azure. I guess we could somehow add logic to the CustomizeDiff function (or somewhere else in the plan process) to fail if these attributes are invalid, but I'm not sure that's the best approach/worth the effort.

Reply:
Already is